packages/wdio-browserstack-service/src/util.ts

```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+            'BUILD_TAG': process.env.BUILD_TAG
+        }
+    }
+    process.env.USER_CONFIG_FOR_REPORTING = JSON.stringify(userConfigForReporting)
+    setCredentialsForCrashReportUpload(options, userConfig)
+}
+
+function processError(error: any, fn: Function, args: any) {
+    log.error(`Error in executing ${fn.name} with args ${args}: ${error}`)
+    let argsString: string
+    try {
+        argsString = JSON.stringify(args)
+    } catch (e) {
+        argsString = util.inspect(args, { depth: 2 })
+    }
+    uploadCrashReport(`Error in executing ${fn.name} with args ${argsString} : ${error}`, error && error.stack)
+}
+
+function o11yErrorHandler(fn: Function) {
+    return function (...args: any) {
+        try {
+            // @ts-ignore
+            const result = fn(...args)
+            if (result instanceof Promise) {
+                return result.catch(error => processError(error, fn, args))
+            }
+            return result
+        } catch (error) {
+            processError(error, fn, args)
+        }
+    }
+}
+
+// https://tugayilik.medium.com/error-handling-via-try-catch-proxy-in-javascript-54116dbf783f
+type ClassType = { new(...args: any[]): any; };
+export function o11yClassErrorHandler<T extends ClassType>(errorClass: T): T {
+    const prototype = errorClass.prototype
+
+    if (Object.getOwnPropertyNames(prototype).length < 2) {
+        return errorClass
+    }
+
+    Object.getOwnPropertyNames(prototype).forEach((methodName) => {
+        const method = prototype[methodName]
+        if (typeof method === 'function' && methodName !== 'constructor') {
+            // In order to preserve this context, need to define like this
+            Object.defineProperty(prototype, methodName, {
+                writable: true,
+                value: function(...args: any) {
+                    try {
+                        const result = method.call(this, ...args)
+                        if (result instanceof Promise) {
+                            return result.catch(error => processError(error, method, args))
+                        }
+                        return result
+
+                    } catch (err) {
+                        processError(err, method, args)
+                    }
+                }
+            })
+        }
+    })
+
+    return errorClass
+}
+
+export async function uploadCrashReport(exception: any, stackTrace: string) {
+    try {
+        if (!credentialsForCrashReportUpload.username || !credentialsForCrashReportUpload.password) {
+            credentialsForCrashReportUpload = process.env.CREDENTIALS_FOR_CRASH_REPORTING !== undefined ? JSON.parse(process.env.CREDENTIALS_FOR_CRASH_REPORTING) : credentialsForCrashReportUpload
+        }
+    } catch (error) {
+        return log.error(`[Crash_Report_Upload] Failed to parse user credentials while reporting crash due to ${error}`)
+    }
+    if (!credentialsForCrashReportUpload.username || !credentialsForCrashReportUpload.password) {
```

Comment:
We have the same `if` in line 216 so we could move this to the top of the function and save some code.

Reply:
The `if` in line 216 runs only once per worker to set the credentials global variable if username & password fields are not present on the current value of this variable or if this variable is an empty JSON as set initially. If fields username & password are not found in the parsed JSON object, we would like to return from this method so an additional check is needed.